---
title: OpenGWAS walkthrough
author: Gibran Hemani
date: 2025-05-08
---

## Background

The objective of OpenGWAS is to reduce the friction between applying statistical methods to GWAS summary data, and to make use of having a large amount of GWAS summary data easily queryable and harmonised in one place.

![OpenGWAS ecosystem](https://gwas.mrcieu.ac.uk/static/images/gwas_ecosystem.bb869525db1a.png)

## Outline

- Installation
- Authenticating to access OpenGWAS
- Performing colocalisation between two traits at a specific locus using the cloud or running locally
- Perform a PheWAS analysis using the OpenGWAS API
- LD matrix operations
- Documentation and vignettes

## Installation

The OpenGWAS ecosystem includes a number of packages. To help with installation they have been pre-compiled on R-universe here: https://mrcieu.r-universe.dev/builds

Let's install the following packages:

- `ieugwasr` - an R interface to the OpenGWAS API
- `gwasvcf` - an R interface to the VCF files that can be analysed locally
- `TwoSampleMR` - an R package for Mendelian randomisation using summary data
- `gwasglue` - an R package for connecting GWAS summary data to other methods

```{r}
install.packages(c('TwoSampleMR', 'ieugwasr', 'gwasglue', 'coloc'), repos = c('https://mrcieu.r-universe.dev', 'https://cloud.r-project.org'))
```

Install gwasvcf using pre-compiled binaries:

```{r}
install.packages("BiocManager", repos = c('https://bioc.r-universe.dev', 'https://cloud.r-project.org'))
BiocManager::install("VariantAnnotation")
install.packages('gwasvcf', repos = c('https://mrcieu.r-universe.dev', 'https://cloud.r-project.org'))
library(gwasvcf)
```

or in the standard way:

```{r}
remotes::install_github("mrcieu/gwasvcf", force=TRUE)
```

Note that docker images are available for some of these packages also - https://hub.docker.com/u/mrcieu

``` bash
docker run --rm -it -v $(pwd):/workspace -w /workspace mrcieu/opengwas-r /bin/bash
```

## Authenticating to access OpenGWAS

1. Visit here: https://api.opengwas.io/
2. Generate a token
3. Save the token in a file called `.Renviron` in your home directory. The file should contain the following line: `OPENGWAS_JWT=your token here`
4. Either restart R or run `readRenviron("~/.Renviron")` 

You can reset your token at any time using the website. Just remember to update your `.Renviron` file with the new token.

## What traits are available?

- Search at [https://gwas.mrcieu.ac.uk/](https://gwas.mrcieu.ac.uk/)
- Use your own GWAS summary data

## Performing colocalisation

For this analysis you need to extract a region of interest from two traits and then run the colocalisation analysis. We'll obtain VCF files and perform the analysis. Instructions here: https://mrcieu.github.io/gwasglue/articles/colocalisation.html

```{r}
library(ieugwasr)
library(gwasvcf)
library(gwasglue)
library(dplyr)
library(coloc)
library(ggplot2)
set_bcftools("bcftools")
readRenviron("env") # Get the OpenGWAS token
```

We'll look at a variant associating with LDL cholesterol levels and coronary heart disease. First download the VCF files:


Get a list of top hits for LDL cholesterol:

```{r}
top <- ieugwasr::tophits('ieu-a-300') %>% arrange(p)
top
```

Let's use the top hit, and extract the region around it. Define the region:

```{r}
chrpos <- paste0(top$chr[1], ":", top$position[1] - 90000, "-", top$position[1] + 90000)
chrpos
```

Extract the summary data for this region from the two traits in the cloud:

```{r}
out <- ieugwasr_to_coloc("ieu-a-300", "ieu-a-7", chrpos)
str(out)
```

Perform the colocalisation analysis:

```{r}
coloc::coloc.abf(out[[1]], out[[2]])
```

```{r}
p <- bind_rows(
    tibble(z = abs(out[[1]]$z), pos = out[[1]]$pos, trait = "LDL cholesterol"),
    tibble(z = abs(out[[2]]$z), pos = out[[2]]$pos, trait = "Coronary heart disease")
) %>%
    ggplot(aes(x = pos, y = z)) +
        geom_point() +
        facet_grid(trait ~ ., scales = "free_y")
ggsave(p, file="ldl_chd2.png", width = 8, height = 8)
```


We can also do this locally by downloading the VCF files. First get links to download the files:

```{r}
links <- gwasinfo_files(c("ieu-a-300", "ieu-a-7"))
links

# Download the files
sapply(unlist(links), function(x) {
    download.file(x, basename(x))
})
```

Extract the data from LDL and CHD, and format it for coloc:

```{r}
vout <- gwasvcf_to_coloc("ieu-a-300.vcf.gz", "ieu-a-7.vcf.gz", chrpos)
str(vout)
```

```{r}
vres <- coloc::coloc.abf(vout[[1]], vout[[2]])
str(vres)
```

Plot the region:

```{r}
p <- bind_rows(
    tibble(z = abs(vout[[1]]$z), pos = vout[[1]]$pos, trait = "LDL cholesterol"),
    tibble(z = abs(vout[[2]]$z), pos = vout[[2]]$pos, trait = "Coronary heart disease")
) %>%
    ggplot(aes(x = pos, y = z)) +
        geom_point() +
        facet_grid(trait ~ ., scales = "free_y")
ggsave(p, file="ldl_chd.png", width = 8, height = 8)
```

## Perform a PheWAS analysis using the OpenGWAS API

A benefit of collating GWAS summary data in one place is that it allows for analysis across all traits. A PheWAS analysis tests a single variant for association against a large number of traits.

Let's use the same variant as above, and test it against all traits in the OpenGWAS database. We'll use the `ieugwasr` package to do this.

```{r}
ldl_phewas <- ieugwasr::phewas(top$rsid[1])
ldl_phewas %>% select(trait, beta, p) %>% slice_head(n=50) %>% as.data.frame()
```

## LD operations

A lot of the analyses that use GWAS summary data require LD information. 

Generate an LD matrix for a set of variants using the 1000G EUR reference panel:

```{r}
ld <- ieugwasr::ld_matrix(top$rsid, pop="EUR")
ld[1:10,1:10]
```

This is an already clumped dataset, so we can see that it actually has very little LD between the variants. Note that LD matrix operations can be performed using a local LD reference panel:

```{r}
ld <- ieugwasr::ld_matrix(top$rsid, bfile="/project/data/reference/ld/EUR", plink_bin="plink")
ld[1:10,1:10]
```

Clumping can also be performed. For example let's read in all SNPs with p < 5e-8 from CHD

```{r}
chd <- gwasvcf::query_gwas("ieu-a-7.vcf.gz", pval=5e-8)
chd 
```

This found 2022 SNPs. Let's clump using a p-val threshold of 1e-6

```{r}
chd_clumped <- clump_gwasvcf(chd, bfile="/project/data/reference/ld/EUR", plink_bin="plink", clump_p = 1e-6)
chd_clumped
```

Note that clumping can be performed on non-VCF files, or on the data within OpenGWAS on the cloud also. See vignettes e.g. https://mrcieu.github.io/ieugwasr/

## Documentation and vignettes

- OpenGWAS API documentation - https://api.opengwas.io/
- Mendelian randomization - https://mrcieu.github.io/TwoSampleMR/
- Linking to other tools - https://mrcieu.github.io/gwasglue/
- Examining pleiotropy - https://explodecomputer.github.io/tryx/
- Multi-ancestry MR and other analyses - https://mrcieu.github.io/CAMERA/

- There is a blog that will be updated with relevant changes and information relating to OpenGWAS - https://opengwas.github.io/
- Note that whenever you load the TwoSampleMR package it will relay a message - this often provides critical recent information.


